mlir/lib/ExecutionEngine/ROCm/BackendUtils.cpp

```diff
@@ -39,19 +39,106 @@
 // TODO: remove this once the rocm_agent_enumerator is ready
 #include "hip/hip_runtime.h"
 
+#include <cstring>
 #include <mutex>
 #include <numeric>
 
 using namespace mlir;
 
 static constexpr const char kRunnerProgram[] = "mlir-rocm-runner";
 static constexpr const char kRocmAgentEnumerator[] = "rocm_agent_enumerator";
+static constexpr const char gcnArchDelimiter[] = ":";
+static constexpr const char targetTriple[] = "amdgcn-amd-amdhsa";
+
+namespace {
+// TODO: remove this once the rocm_agent_enumerator is ready
+void getGpuGCNArchName(hipDevice_t device, std::string &gcnArchName) {
+  hipDeviceProp_t props;
+  hipError_t result = hipGetDeviceProperties(&props, device);
+  if (result != hipSuccess) {
+    gcnArchName = "";
+    return;
+  }
+
+  const char *pArchName = props.gcnArchName;
+  gcnArchName.assign(pArchName);
+}
+
+// This function convert the arch name to target features:
+// gfx908:sramecc+:xnack- to +sramecc,-xnack
+std::string getFeaturesFromArchName(const std::string &gcnArchName) {
+  std::string features;
```

Comment:
Maybe we should use a https://llvm.org/docs/ProgrammersManual.html#the-twine-class here - and thus avoid a bunch of copies and the like

Reply:
(that is, instead of what we've got now, we accumulate a bunch of "this memory from here to here" references)

On the other hand, that's probably overengineering for this use-case - I just wanted to point out the possibility.

Reply:
I can see `concat` on `Twine` would be a more efficient move on the 3rd step but I'm not sure if step 1/2 can get additional benefit over the use of twine. For the 3rd step, I can also use everything as-is but accumulate into a lambda captured `std::string` and avoid the temporaries. As you have mentioned, if this is a token string with thousands of elements in them we sure have better ways to handle it. But they really wouldn't make much difference at the use case we have here.

Reply:
Yeah - there's no point in really over-engineering this. I just figure it's worth thinking about obvious efficiency issues and considering if the clarity is worth the tradeoff